ur_controllers/src/ur_configuration_controller.cpp

```diff
@@ -0,0 +1,122 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+//----------------------------------------------------------------------
+/*!\file
+ *
+ * \author  Jacob Larsen jala@universal-robots.com
+ * \date    2024-07-11
+ *
+ *
+ *
+ *
+ */
+//----------------------------------------------------------------------
+
+#include <ur_controllers/ur_configuration_controller.hpp>
+namespace ur_controllers
+{
+
+controller_interface::CallbackReturn URConfigurationController::on_init()
+{
+  param_listener_ = std::make_shared<ur_configuration_controller::ParamListener>(get_node());
+  params_ = param_listener_->get_params();
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn
+URConfigurationController::on_configure(const rclcpp_lifecycle::State& /* previous_state */)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::InterfaceConfiguration URConfigurationController::command_interface_configuration() const
+{
+  // No command interfaces currently
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  const std::string tf_prefix = params_.tf_prefix;
+
+  return config;
+}
+
+controller_interface::InterfaceConfiguration URConfigurationController::state_interface_configuration() const
+{
+  controller_interface::InterfaceConfiguration config;
+  config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+
+  const std::string tf_prefix = params_.tf_prefix;
+
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_major");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_minor");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_build");
+  config.names.emplace_back(tf_prefix + "get_robot_software_version/get_version_bugfix");
+
+  return config;
+}
+
+controller_interface::return_type URConfigurationController::update(const rclcpp::Time& /* time */,
+                                                                    const rclcpp::Duration& /* period */)
+{
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::CallbackReturn
+URConfigurationController::on_activate(const rclcpp_lifecycle::State& /* previous_state */)
+{
+  get_robot_software_version_srv_ = get_node()->create_service<ur_msgs::srv::GetRobotSoftwareVersion>(
+      "~/get_robot_software_version", std::bind(&URConfigurationController::getRobotSoftwareVersion, this,
+                                                std::placeholders::_1, std::placeholders::_2));
```

Comment:
This should go to `on_configure()`. `on_activate()` is supposed to be real-time safe which creating a service definitively isn't.